File: src/parsing/parser.rs
#![allow(dead_code, non_snake_case, unused_variables, unused)]

use crate::lexerization::{Token, TokenType};
use crate::lexerization::TokenType::NUMBER;
use crate::parsing::{BlockNode, GeneralRegNode, InvalidNode, MoveCommandNode, NumberNode, NumberType, SyscallArgNode, SyscallCommandNode, SyscallIdNode};
use crate::parsing::errors::UnknownCommandError;
use crate::parsing::nodes::ASTNode;

pub struct Parser {
    input: Vec<Vec<Token>>,
    line_position: usize,

    pub top_layer_nodes: Vec<Box<dyn ASTNode>>
}

impl Parser {
    pub fn new(input: Vec<Vec<Token>>) -> Self
    {
        Self {
            input,
            line_position: 0,
            top_layer_nodes: Vec::new()
        }
    }

    pub fn parse(&mut self) -> &Vec<Box<dyn ASTNode>>
    {
        let mut for_index: usize = 0;

        for line in self.input.clone() {
            if for_index < self.line_position { for_index += 1; continue }

            for token in line {
                match token.token_type {
                    TokenType::KW_BLOCK => {
                        let node = Box::new(self.parseBlock());

                        self.top_layer_nodes.push(node)
                    }
                    TokenType::KW_MACRO => {

                    }

                    _ => {
                        // error
                    }
                }
            }

            self.line_position += 1;
            for_index += 1;
        }

        &self.top_layer_nodes
    }

    fn parseBlock(&mut self) -> BlockNode {
        let mut name = String::new();
        let mut nodes_in: Vec<Box<dyn ASTNode>> = Vec::new();

        let mut for_index: usize = 0;

        'LINE_MARK: for line in self.input.clone() {
            if for_index < self.line_position {
                for_index += 1;
                continue
            }

            let mut is_block_before = false;
            let mut is_name_before = false;

            for token in line {
                if token.token_type == TokenType::KW_BLOCK {
                    is_block_before = true;
                }
                if token.token_type == TokenType::IDENTIFIER && is_block_before {
                    name = token.value;
                    is_block_before = false;
                    is_name_before = true;
                }
                if token.token_type == TokenType::START_CODE_SPACE && is_name_before {
                    self.line_position += 1;
                    for_index += 1;
                    is_name_before = false;

                    match self.parseCommands() {
                        Ok(ok) => {nodes_in = ok}
                        Err(e) => {
                            println!("{}", e);

                            break 'LINE_MARK;
                        }
                    }

                    break 'LINE_MARK
                }
            }

            self.line_position += 1;
            for_index += 1;
        }

        BlockNode::new(name, nodes_in)
    }

    fn parseCommands(&mut self) -> Result<Vec<Box<dyn ASTNode>>, UnknownCommandError> {
        let mut nodes: Vec<Box<dyn ASTNode>> = Vec::new();

        let mut for_index: usize = 0;

        let lines = self.input.clone();

        for line in lines {
            if for_index < self.line_position { for_index += 1; continue }

            match &line[0].token_type {
                TokenType::CMD_MOVE => {
                    let i= self.parseMoveCommand();
                    nodes.push(Box::new(i))
                }
                TokenType::CMD_SYSCALL => {
                    nodes.push(Box::new(SyscallCommandNode::new()))
                }

                TokenType::END_CODE_SPACE => {
                    break
                }

                _e => {
                    return Err(
                        UnknownCommandError::new(
                            String::from("main.plmr.pasm"),
                            self.line_position,
                            line[0].value.clone()
                        )
                    )
                }
            }

            for_index += 1;
            self.line_position += 1;
        }

        Ok(nodes)
    }

    fn parseMoveCommand(&mut self) -> MoveCommandNode {
        let mut node_receiver: Box<dyn ASTNode> = Box::new(InvalidNode::new());
        let mut node_value: Box<dyn ASTNode> = Box::new(InvalidNode::new());

        let mut for_index: usize = 0;

        for line in self.input.clone() {
            if for_index < self.line_position { for_index += 1; continue }

            if !(line[0].token_type == TokenType::CMD_MOVE || line[1].token_type == TokenType::CMD_AND_ARGS_DIVIDER) {
                // error
                break
            }

            let args_token: Vec<Token> = line
                .into_iter()
                .skip_while(|token| token.token_type != TokenType::CMD_AND_ARGS_DIVIDER)
                .skip(1)
                .collect();

            let mut iter = args_token.split(|token| token.token_type == TokenType::ARGS_DIVIDER);

            let arg1_tkn = iter.next().unwrap_or(&[]).to_vec();
            let arg2_tkn = iter.next().unwrap_or(&[]).to_vec();

            node_receiver = self.parseExpression(arg1_tkn);
            node_value = self.parseExpression(arg2_tkn);

            break
        };


        MoveCommandNode::new(node_receiver, node_value)
    }

    fn parseExpression(&mut self, expression: Vec<Token>) -> Box<dyn ASTNode> {
        if expression[0].token_type == TokenType::VAR_PREFIX || expression[0].token_type == TokenType::REG_PREFIX {
            if expression[1].token_type == TokenType::IDENTIFIER {
                match expression[1].value.as_str() {
                    "syscall_id" => {
                        return Box::new(SyscallIdNode::new())
                    }

                    _ => {
                        if expression[1].value.starts_with("syscall_arg") {
                            if let Ok(num) = expression[1].value["syscall_arg".len()..].parse::<u8>() {
                                return Box::new(SyscallArgNode::new(num));
                            }
                        }
                        if expression[1].value.starts_with("general_reg") {
                            if let Ok(num) = expression[1].value["general_reg".len()..].parse::<u8>() {
                                return Box::new(GeneralRegNode::new(num));
                            }
                        }
                    }
                }
            } else {
                // err
            }
        }
        if expression[0].token_type == NUMBER && expression.len() == 1 {
            return if expression[0].value.contains('.') {
                Box::new(NumberNode::new(expression[0].clone().value, NumberType::FLOAT))
            } else {
                Box::new(NumberNode::new(expression[0].clone().value, NumberType::INTEGER))
            }
        }

        Box::new(InvalidNode::new())
    }
}

File: src/main/kotlin/parsing/Parser.kt
package parsing

import lexicalization.DToken
import lexicalization.ETokenType
import parsing.nodes.AASTNode
import parsing.nodes.base.DBlockNode
import parsing.nodes.base.DPermanentDataNode
import parsing.nodes.commands.DMoveCommandNode
import parsing.nodes.commands.DSystemCallCommandNode
import parsing.nodes.expressions.DIdNode
import parsing.nodes.expressions.DIntegerNode
import parsing.nodes.expressions.DMinusNode
import parsing.nodes.expressions.DStringNode
import parsing.nodes.regs_and_vars.DCurrentAddressVarNode
import parsing.nodes.regs_and_vars.DGeneralRegNode
import parsing.nodes.regs_and_vars.system_calls.DSystemCallArgVarNode
import parsing.nodes.regs_and_vars.system_calls.DSystemCallIdVarNode
import utils.errors.DParserError
import java.lang.reflect.Type
import kotlin.reflect.KClass

class Parser(private val input: List<List<DToken>>) {
    private val parsed: MutableList<AASTNode> = mutableListOf()
    private var current_line_index = 0

    fun parse(): MutableList<AASTNode> {
        for(line in input.withIndex()){
            if(line.index < current_line_index) continue

            if (line.value[0].type == ETokenType.KW_BLOCK &&
                line.value[1].type == ETokenType.IDENTIFIER &&
                line.value[2].type == ETokenType.START_CODE_SPACE){
                parsed += parseBlock(current_line_index)
            }
            else if (line.value[0].type == ETokenType.KW_PERMANENT)
                parsePermanent()
            else
                throw DParserError("Unknown top-layer construction on line ${line.index + 1}")
        }

        return parsed
    }

    private fun parseBlock(start_index: Int): DBlockNode {
        var name = ""
        val children: MutableList<AASTNode> = mutableListOf()

        for(line in input.withIndex()){
            if(line.index < current_line_index) continue

            if(line.index == start_index) {
                name = line.value[1].value!!
            }
            else if(line.value[0].type == ETokenType.END_CODE_SPACE) {
                current_line_index++
                break
            }
            else {
                current_line_index++

                children += parseCommand()
            }
        }

        if( name.isBlank() || children.isEmpty() ) throw DParserError("Empty name or no commands in block at ${start_index + 1}th line")

        current_line_index++

        return DBlockNode(name, children)
    }

    private fun parsePermanent(): DPermanentDataNode {
        val line = input[current_line_index]

        if( line[0].type != ETokenType.KW_PERMANENT
            || line[1].type != ETokenType.IDENTIFIER
            || line[2].type != ETokenType.CMD_AND_ARGS_DIVIDER ) throw DParserError("Invalid permanent value at ${current_line_index} line")

        current_line_index++

        return DPermanentDataNode(line[1].value!!, parseExpression(line.slice(3..<line.size)))
    }

    private fun parseCommand(): AASTNode {
        return when(input[current_line_index][0].type){
            ETokenType.CMD_SYSTEM_CALL -> DSystemCallCommandNode()
            ETokenType.CMD_MOVE -> parseMoveCommand()

            else -> throw DParserError("Unknown command at $current_line_index, command: ${input[current_line_index]}")
        }
    }

    private fun parseMoveCommand(): DMoveCommandNode{
        val arg1_tkn = mutableListOf<DToken>()
        val arg2_tkn = mutableListOf<DToken>()

        var first_flag = false
        var second_flag = false

        for(token in input[current_line_index]) {
            if(token.type == ETokenType.CMD_AND_ARGS_DIVIDER) { first_flag = true; continue }
            if (first_flag){
                if(token.type == ETokenType.ARGS_DIVIDER){
                    first_flag = false
                    second_flag = true
                    continue
                }
                else {
                    arg1_tkn += token
                }
            }
            if(second_flag) {
                arg2_tkn += token
            }
        }


        return DMoveCommandNode(
            receiver = parseRegOrVar(arg1_tkn),
            value = parseExpression(arg2_tkn)
        )
    }

    private fun parseSimpleExpression(from: DToken) = when (from.type) {
        ETokenType.NUMBER -> DIntegerNode(from.value!!.toInt())
        ETokenType.STRING -> DStringNode(from.value!!)
        ETokenType.IDENTIFIER -> DIdNode(from.value!!)

        else -> throw DParserError("Invalid expression at $from")
    }

    private fun parseExpression(from: List<DToken>) = when {
        from.size == 1 -> parseSimpleExpression(from[0])

        from[0].type == ETokenType.KW_TERM -> {
            lateinit var part_first: AASTNode
            lateinit var part_second: AASTNode
            lateinit var symbol: ETokenType
            var first_or_second = false


            for ( token in from.slice(1..<from.size).withIndex() ) {
                when(token.value.type) {
                    ETokenType.VAR_PREFIX, ETokenType.REG_PREFIX ->
                        if( first_or_second )
                            part_first = parseRegOrVar(from.slice(token.index + 1..token.index + 2))
                        else
                            part_second = parseRegOrVar(from.slice(token.index + 1..token.index + 2))

                    ETokenType.MINUS, ETokenType.PLUS, ETokenType.MULTIPLY, ETokenType.DIVIDE -> {
                        symbol = token.value.type
                        first_or_second = !first_or_second
                    }

                    ETokenType.NUMBER, ETokenType.STRING, ETokenType.IDENTIFIER ->
                        if( first_or_second )
                            part_first = parseSimpleExpression(token.value)
                        else
                            part_second = parseSimpleExpression(token.value)



                    else ->
                        throw DParserError("Invalid expression at $from")

                }
            }

            when(symbol) {
                ETokenType.MINUS -> DMinusNode (
                    part_first,
                    part_second
                )

                else -> throw DParserError("Invalid expression at $from")
            }

        }

        else -> throw DParserError("Invalid expression at $from")
    }

    private fun parseRegOrVar(from: List<DToken>): AASTNode = when (from[0].type) {
        ETokenType.REG_PREFIX ->
            when {
                from[1].value!!.startsWith("general_reg") -> DGeneralRegNode(from[1].value!!)

                else -> throw DParserError("Undefined reg or var: $from")
            }
        ETokenType.VAR_PREFIX ->
            when {
                from[1].value!! == "system_call_id" -> DSystemCallIdVarNode()
                from[1].value!! == "current_address" -> DCurrentAddressVarNode()
                from[1].value!!.startsWith("system_call_arg") -> DSystemCallArgVarNode(from[1].value!!)

                else -> throw DParserError("Undefined reg or var: $from")
            }

        else -> throw DParserError("Undefined reg or var: $from")
    }


}